lib/proxy/peer/quicserver.go

```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+		log.WarnContext(c.Context(),
+			"Dial request has out of sync timestamp, 0-RTT performance will be impacted.",
+			"request_timestamp", requestTimestamp,
+		)
+		select {
+		case <-c.HandshakeComplete():
+		case <-c.Context().Done():
+			return
+		}
+	}
+
+	// a replayed request is always wrong even after a full handshake, the
+	// replay might've happened before the legitimate request
+	if !s.replayStore.add(req.GetNonce(), time.Now()) {
+		log.ErrorContext(c.Context(), "Request is reusing a nonce, rejecting.", "nonce", req.GetNonce())
+		sendErr(trace.BadParameter("reused or invalid nonce"))
+		return
+	}
+
+	_, clusterName, ok := strings.Cut(req.GetServerId(), ".")
+	if !ok {
+		sendErr(trace.BadParameter("server_id %q is missing cluster information", req.GetServerId()))
+		return
+	}
+
+	nodeConn, err := s.clusterDialer.Dial(clusterName, DialParams{
+		From: &utils.NetAddr{
+			Addr:        req.GetSource().GetAddr(),
+			AddrNetwork: req.GetSource().GetNetwork(),
+		},
+		To: &utils.NetAddr{
+			Addr:        req.GetDestination().GetAddr(),
+			AddrNetwork: req.GetDestination().GetNetwork(),
+		},
+		ServerID: req.GetServerId(),
+		ConnType: types.TunnelType(req.GetConnType()),
+	})
+	if err != nil {
+		sendErr(err)
+		return
+	}
+	defer nodeConn.Close()
+
+	var eg errgroup.Group
+	eg.Go(func() error {
+		defer st.Close()
+		// an empty protobuf message has an empty wire encoding, so by sending a
+		// size of 0 (i.e. four zero bytes) we are sending an empty DialResponse
+		// with an empty Status, which signifies a successful dial
+		if _, err := st.Write(binary.LittleEndian.AppendUint32(nil, 0)); err != nil {
+			return trace.Wrap(err)
+		}
+		_, err := io.Copy(st, nodeConn)
+		return trace.Wrap(err)
+	})
+	eg.Go(func() error {
+		defer st.CancelRead(0)
+
+		// wait for the handshake before forwarding application data from the
+		// client; the client shouldn't be sending application data as 0-RTT
+		// anyway, but just in case
+		select {
+		case <-c.HandshakeComplete():
+		case <-c.Context().Done():
+			return trace.Wrap(context.Cause(c.Context()))
+		}
+
+		_, err := io.Copy(nodeConn, st)
+		return trace.Wrap(err)
+	})
+	_ = eg.Wait()
 }
 
 // Close stops listening for incoming connections and ungracefully terminates
 // all the existing ones.
 func (s *QUICServer) Close() error {
-	panic("QUIC proxy peering is not implemented")
+	s.runCancel()
+	s.Shutdown(context.Background())
+	return nil
 }
 
 // Shutdown stops listening for incoming connections and waits until the
 // existing ones are closed or until the context expires. If the context
 // expires, running connections are ungracefully terminated.
 func (s *QUICServer) Shutdown(ctx context.Context) error {
-	panic("QUIC proxy peering is not implemented")
+	s.mu.Lock()
+	s.closed = true
+	s.mu.Unlock()
+
+	defer s.runCancel()
+	defer context.AfterFunc(ctx, s.runCancel)()
+	s.serveCancel()
+	s.wg.Wait()
```

Comment:
```suggestion
	defer context.AfterFunc(ctx, s.runCancel)()
	s.serveCancel()
	s.wg.Wait()
	s.runCancel()
```

Reply:
I've been liking the pattern of
```go
defer idempotentFoo()
defer context.AfterFunc(ctx, idempotentFoo)()
```

to cancel the `AfterFunc` because we're obviously going to run the callback ourselves - which also guarantees that we're not leaving the callback attached to the context.

Reply:
It's a subtlety that might be lost in most people looking at this. Works fine so I'll resolve.